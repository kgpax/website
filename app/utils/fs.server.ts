export { default } from 'fs/promises'
