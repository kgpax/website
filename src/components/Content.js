import styled from 'styled-components';
import { space } from 'styled-system';
import PropTypes from 'prop-types';

const StyledContent = styled.article`
  width: 100%;
  max-width: 65rem;
  margin: 0 auto;
  text-align: ${({ align }) => align};
  ${space}

  @media only screen and (min-width: 768px) {
    max-width: 70rem;
  }
`;

const Content = props => <StyledContent {...props} />;
Content.propTypes = {
  align: PropTypes.string,
};
Content.defaultProps = {
  align: 'left',
};

export default Content;
