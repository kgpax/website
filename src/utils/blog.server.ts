import glob from 'fast-glob'
import {
  BlogData,
  BlogEntry,
  BlogSeriesEntry,
  EntryMetaData,
} from '~/types/blog'

export async function getBlogEntries(): Promise<BlogEntry[]> {
  const postPaths = await glob('src/pages/blog/**/*.mdx')

  const posts = await Promise.all(
    postPaths.map(async path => {
      const filePath = path.replace(/^src\/pages/, '')
      const fileName = path.split('/').reverse()[0].replace(/\.mdx/, '')
      const slug = filePath.replace(/\.mdx/, '')
      const series = filePath
        .substring(0, filePath.lastIndexOf('/'))
        .replace(/^\/blog\//, '')
      const { meta }: { meta: EntryMetaData } = await import(
        `../pages${filePath}`
      )

      return {
        ...meta,
        filePath,
        fileName,
        series,
        slug,
        isPublished: !!meta.date,
      } as BlogData
    }),
  )

  const sortedPosts = posts.sort((a, b) =>
    (a.date ?? '9999-99-99').localeCompare(b.date ?? '9999-99-99'),
  )

  const entries: BlogEntry[] = []

  for (const post of sortedPosts) {
    if (!post.series) {
      entries.push({
        type: 'post',
        ...post,
      })
    } else {
      if (post.series) {
        let series = entries.find(
          entry => entry.series === post.series,
        ) as BlogSeriesEntry
        if (!series) {
          series = {
            type: 'series',
            ...post,
            posts: [],
          }
          entries.push(series)
        }
        if (post.fileName !== '_meta') {
          series.posts.push({ type: 'post', ...post })
        }
      } else {
      }
    }
  }

  return entries
}
