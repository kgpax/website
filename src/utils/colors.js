const pairs = [
  ['#D7F2BA', '#BDE4A8'],
  ['#FFF200', '#FFD300'],
  ['#97D2FB', '#83BCFF'],
  ['#FFBE85', '#FF9B42'],
];

export const pickColorPair = () =>
  pairs[Math.floor(Math.random() * pairs.length)];
