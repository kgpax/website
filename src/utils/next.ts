import { GetStaticPropsResult } from 'next'

export function next404<T>(): GetStaticPropsResult<T> {
  return { notFound: true }
}
