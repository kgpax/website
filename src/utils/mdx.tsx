import { MDXComponents } from 'mdx/types'
import { BlogPage, Heading, Link, MdxCode } from '~/components'

export const components: MDXComponents = {
  wrapper: props => {
    console.log(props)
    return <BlogPage {...props}>{props.children}</BlogPage>
  },
  h2: Heading,
  a: Link,
  code: MdxCode,
}
