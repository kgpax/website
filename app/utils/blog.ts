import parseFrontMatter from 'front-matter'
import { BlogInfo, BlogMetadata, BlogMetadataAndUrl } from '~/types/blog'
import fs from '~/utils/fs.server'
import { bundleMDX } from '~/utils/mdx.server'
import process from '~/utils/process.server'

const blogContentPath = (filename: string = '') =>
  `${process.cwd()}/app/content/blog/${filename}`

export async function getPost(slug: string): Promise<BlogInfo | null> {
  const source = await fs
    .readFile(blogContentPath(`${slug}.mdx`), 'utf-8')
    .catch(() => null)

  if (source === null) return null

  const post = await bundleMDX({
    source,
    mdxOptions(options) {
      options.remarkPlugins = [...(options.remarkPlugins ?? [])]
      options.rehypePlugins = [...(options.rehypePlugins ?? [])]
      return options
    },
    esbuildOptions(options) {
      options.platform = 'node'
      return options
    },
    cwd: blogContentPath(),
  })

  return {
    code: post.code,
    url: `/blog/${slug}`,
    ...(post.frontmatter as BlogMetadata),
  }
}

export async function getPosts(): Promise<BlogMetadataAndUrl[]> {
  const postsPath = await fs.readdir(blogContentPath(), {
    withFileTypes: true,
  })

  const posts = await Promise.all(
    postsPath.map(async entry => {
      const file = await fs.readFile(blogContentPath(entry.name))
      const { attributes } = parseFrontMatter<BlogMetadata>(file.toString())

      return {
        url: `/blog/${entry.name.replace(/\.mdx/, '')}`,
        ...attributes,
      } as BlogMetadataAndUrl
    }),
  )

  return posts
}
