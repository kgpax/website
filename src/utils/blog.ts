import { BlogSeriesEntry } from '~/types/blog'

export function getBlogPostSeries(
  series: BlogSeriesEntry[],
  id: string,
): BlogSeriesEntry | null {
  return series.find(x => x.posts.some(y => y.id === id)) ?? null
}
