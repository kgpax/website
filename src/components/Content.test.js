import { render } from '@test/utils';
import Content from '@components/Content';

const component = (props = {}) =>
  render(<Content {...props}>This is some content</Content>);

test('should match snapshot', () => {
  expect(component().element).toMatchSnapshot();
});

test('should be an ARTICLE element', () => {
  const { element } = component();
  expect(element.tagName).toBe('ARTICLE');
});

test('should contain expected content', () => {
  const { element } = component();
  expect(element.textContent).toBe('This is some content');
});

test('should default to left text alignment', () => {
  const { element } = component();
  expect(element).toHaveStyle({ 'text-align': 'left' });
});

test('should allow text alignment to be set', () => {
  const { element } = component({ align: 'right' });
  expect(element).toHaveStyle({ 'text-align': 'right' });
});
