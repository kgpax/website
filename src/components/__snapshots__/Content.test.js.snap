// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<article
  class="Content__StyledContent-sc-18ob92m-0 lodTZg"
>
  This is some content
</article>
`;
